### ISSUE TITLE
Formatting User-Defined Types

### ISSUE THREAD

**Issue by ywqzzy** [pingcap/tiflash#5880]
## Enhancement
In TiFlash we use fmt library to serialize a bunch of objects, but we don't support formatting User-defined types.
We can implement format method for our user-defined types by specializing the formatter<T> struct template.
Can ref to https://fmt.dev/latest/api.html?highlight=fmt%20formatter#formatting-user-defined-types.

**Comment by ywqzzy** [comment]
Many examples can be found in Bustub. https://github.com/cmu-db/bustub/blob/master/src/include/binder/bound_table_ref.h

**Comment by JaySon-Huang** [comment]
Maybe you can list the structs you want to support formatting in this issue?

**Comment by ywqzzy** [comment]
> Maybe you can list the structs you want to support formatting in this issue?


1. MPPTaskId
2. All physicalPlan
3. ExpressionAction
4. RegionInfo
5. DecodeDetail
6. ServerInfo
7. FmtBuffer
etc...

**Comment by Lloyd-Pottiger** [comment]
Any advantages comparing to define a `toString` function?

**Comment by ywqzzy** [comment]
> For enum type, we can use magic_enum later, for more complex type, we can define a `toString` function. Any advantages for implementing format method like this?
`toString()` functions are necessary for fmt::formatting.
We can use the syntax like 
```
fmt::format("{}", user_defined_object);
```
instead of
```
fmt::format("{}", user_defined_object.toString());
```
if we specialized the formatter struct template.

Anyway, I think we should provide the ability to log easier, and it will make debug easier for future

**Comment by Lloyd-Pottiger** [comment]
yeah, but I think `toString` can also satisfy the demand, even I think they are almost the same, I do not find any advantages about it. Agree to use this method in later development, but seems not need to refine the existed code.

**Comment by fuzhe1989** [comment]
It could reduce some object constructions compared to `toString`.
